File: stdlib/js/hull/middleware/etag.js
/*
 * hull:middleware:etag -- ETag response helpers
 *
 * Wraps res.json/text/html to add ETag support with 304 Not Modified.
 * Provides wrapper functions that replace res.json() in route handlers.
 *
 * Usage:
 *   import { etag } from "hull:middleware:etag";
 *   app.get("/api/items", (req, res) => {
 *       const data = db.query("SELECT * FROM items");
 *       etag.json(req, res, data);
 *   });
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { crypto } from "hull:crypto";

const MAX_BODY_SIZE = 1024 * 1024; // 1 MB

function compute(body) {
    if (!body || body.length === 0) return null;
    const hash = crypto.sha256(body);
    return 'W/"' + hash.substring(0, 16) + '"';
}

function matches(req, tag) {
    if (!tag) return false;
    const inm = req.header("If-None-Match");
    if (!inm) return false;
    if (inm.trim() === "*") return true;
    const parts = inm.split(",");
    for (let i = 0; i < parts.length; i++) {
        if (parts[i].trim() === tag) return true;
    }
    return false;
}

function jsonEtag(req, res, data, status) {
    if (req.method !== "GET" && req.method !== "HEAD") {
        if (status) res.status(status);
        res.json(data);
        return;
    }

    const body = JSON.stringify(data);
    if (body.length > MAX_BODY_SIZE) {
        if (status) res.status(status);
        res.json(data);
        return;
    }

    const tag = compute(body);
    if (matches(req, tag)) {
        res.status(304).header("ETag", tag);
        return;
    }

    res.header("ETag", tag);
    if (status) res.status(status);
    res.json(data);
}

function textEtag(req, res, text, status) {
    if (req.method !== "GET" && req.method !== "HEAD") {
        if (status) res.status(status);
        res.text(text);
        return;
    }

    if (text.length > MAX_BODY_SIZE) {
        if (status) res.status(status);
        res.text(text);
        return;
    }

    const tag = compute(text);
    if (matches(req, tag)) {
        res.status(304).header("ETag", tag);
        return;
    }

    res.header("ETag", tag);
    if (status) res.status(status);
    res.text(text);
}

function htmlEtag(req, res, html, status) {
    if (req.method !== "GET" && req.method !== "HEAD") {
        if (status) res.status(status);
        res.html(html);
        return;
    }

    if (html.length > MAX_BODY_SIZE) {
        if (status) res.status(status);
        res.html(html);
        return;
    }

    const tag = compute(html);
    if (matches(req, tag)) {
        res.status(304).header("ETag", tag);
        return;
    }

    res.header("ETag", tag);
    if (status) res.status(status);
    res.html(html);
}

const etag = { compute, matches, json: jsonEtag, text: textEtag, html: htmlEtag };
export { etag };

File: stdlib/lua/hull/middleware/etag.lua
--
-- hull.middleware.etag -- ETag response helpers
--
-- Wraps res:json/text/html to add ETag support with 304 Not Modified.
-- Not a middleware function (runs after handler, not before) — instead
-- provides wrapper functions that replace res:json() in route handlers.
--
-- Usage:
--   local etag = require("hull.middleware.etag")
--   app.get("/api/items", function(req, res)
--       local data = db.query("SELECT * FROM items")
--       etag.json(req, res, data)
--   end)
--
-- SPDX-License-Identifier: AGPL-3.0-or-later
--

local etag = {}

local MAX_BODY_SIZE = 1024 * 1024  -- 1 MB: skip ETag for larger responses

--- Compute a weak ETag from a body string.
-- Returns W/"<first 16 hex chars of SHA-256>"
function etag.compute(body)
    if not body or #body == 0 then return nil end
    local hash = crypto.sha256(body)
    return 'W/"' .. hash:sub(1, 16) .. '"'
end

--- Check if the request's If-None-Match header matches an ETag value.
-- Handles comma-separated lists and * wildcard.
function etag.matches(req, tag)
    if not tag then return false end
    local inm = req.headers["if-none-match"]
    if not inm then return false end

    -- Wildcard
    if inm:match("^%s*%*%s*$") then return true end

    -- Check each comma-separated value
    for part in inm:gmatch("[^,]+") do
        local trimmed = part:match("^%s*(.-)%s*$")
        if trimmed == tag then return true end
    end
    return false
end

--- Send a JSON response with ETag support.
-- If If-None-Match matches, sends 304 (no body).
-- Otherwise sends full response with ETag header.
-- status: optional HTTP status code (default 200)
function etag.json(req, res, data, status)
    -- Only compute ETag for GET/HEAD
    if req.method ~= "GET" and req.method ~= "HEAD" then
        if status then res:status(status) end
        res:json(data)
        return
    end

    local body = json.encode(data)
    if #body > MAX_BODY_SIZE then
        if status then res:status(status) end
        res:json(data)
        return
    end

    local tag = etag.compute(body)
    if etag.matches(req, tag) then
        res:status(304):header("ETag", tag)
        return
    end

    res:header("ETag", tag)
    if status then res:status(status) end
    res:json(data)
end

--- Send a text response with ETag support.
function etag.text(req, res, text, status)
    if req.method ~= "GET" and req.method ~= "HEAD" then
        if status then res:status(status) end
        res:text(text)
        return
    end

    if #text > MAX_BODY_SIZE then
        if status then res:status(status) end
        res:text(text)
        return
    end

    local tag = etag.compute(text)
    if etag.matches(req, tag) then
        res:status(304):header("ETag", tag)
        return
    end

    res:header("ETag", tag)
    if status then res:status(status) end
    res:text(text)
end

--- Send an HTML response with ETag support.
function etag.html(req, res, html, status)
    if req.method ~= "GET" and req.method ~= "HEAD" then
        if status then res:status(status) end
        res:html(html)
        return
    end

    if #html > MAX_BODY_SIZE then
        if status then res:status(status) end
        res:html(html)
        return
    end

    local tag = etag.compute(html)
    if etag.matches(req, tag) then
        res:status(304):header("ETag", tag)
        return
    end

    res:header("ETag", tag)
    if status then res:status(status) end
    res:html(html)
end

return etag
